File: src/main/java/org/lanternpowered/server/util/UUIDHelper.java
/*
 * Lantern
 *
 * Copyright (c) LanternPowered <https://www.lanternpowered.org>
 * Copyright (c) SpongePowered <https://www.spongepowered.org>
 * Copyright (c) contributors
 *
 * This work is licensed under the terms of the MIT License (MIT). For
 * a copy, see 'LICENSE.txt' or <https://opensource.org/licenses/MIT>.
 */
package org.lanternpowered.server.util;

import static com.google.common.base.Preconditions.checkArgument;
import static com.google.common.base.Preconditions.checkNotNull;

import java.util.UUID;

public final class UUIDHelper {

    private UUIDHelper() {
    }

    /**
     * Parses the uuid instance from a flat string (without dashes).
     * 
     * @param string the flat string
     * @return the uuid
     */
    public static UUID fromFlatString(String string) {
        checkNotNull(string, "string");
        checkArgument(string.length() == 32, "length must be 32");
        return UUID.fromString(string.substring(0, 8) + "-" + string.substring(8, 12) + "-" + string.substring(12, 16) +
                "-" + string.substring(16, 20) + "-" + string.substring(20, 32));
    }

    /**
     * Converts the uuid to a flat string (without dashes).
     * 
     * @param uuid the uuid
     * @return the flat string
     */
    public static String toFlatString(UUID uuid) {
        return checkNotNull(uuid, "uuid").toString().replace("-", "");
    }

}

File: src/main/kotlin/org/lanternpowered/server/util/UUIDHelper.kt
/*
 * Lantern
 *
 * Copyright (c) LanternPowered <https://www.lanternpowered.org>
 * Copyright (c) SpongePowered <https://www.spongepowered.org>
 * Copyright (c) contributors
 *
 * This work is licensed under the terms of the MIT License (MIT). For
 * a copy, see 'LICENSE.txt' or <https://opensource.org/licenses/MIT>.
 */
package org.lanternpowered.server.util

import java.util.UUID

object UUIDHelper {

    /**
     * A regex to match [UUID]s.
     */
    val regex: Regex = "[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}".toRegex(RegexOption.IGNORE_CASE)

    /**
     * Attempts to parse the given [value] as an [UUID].
     */
    fun parseOrNull(value: String): UUID? {
        return try {
            UUID.fromString(value)
        } catch (ex: Exception) {
            null
        }
    }

    /**
     * Parses the [UUID] from a flat string (without dashes).
     *
     * @param flat The flat string
     * @return The uuid
     */
    @JvmStatic
    fun parseFlatString(flat: String): UUID {
        check(flat.length == 32) { "length must be 32" }
        val most = java.lang.Long.parseLong(flat, 0, 16, 16)
        val least = java.lang.Long.parseLong(flat, 16, 32, 16)
        return UUID(most, least)
    }

    /**
     * Converts the [UUID] to a flat string (without dashes).
     *
     * @param uuid The uuid
     * @return The flat string
     */
    @JvmStatic
    fun toFlatString(uuid: UUID): String {
        val most = uuid.mostSignificantBits
        val least = uuid.leastSignificantBits
        fun Long.toPart(): String {
            val s = this.toString(16)
            // Guarantee that the length is 16 characters
            return if (s.length != 16) "0".repeat(16 - s.length) + s else s
        }
        return most.toPart() + least.toPart()
    }
}
